### PR TITLE
Added JobTemplate, a preliminary step for ScheduledJob and Workflow

### PR DESCRIPTION
**Pull request by soltysh**
@sdminonne as promised, sorry it took this long :blush: 
@erictune fyi though it does not have to be in for 1.2

---

This change is [Reviewable](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/21675)

### CODE REVIEW

**Comment by sdminonne** [comment]
@soltysh thanks.

File: pkg/apis/batch/v1/types.go [hunk 1]
```diff
@@ -48,6 +48,24 @@ type JobList struct {
 	Items []Job `json:"items"`
 }
 
+// JobTemplate describes a template for creating copies of a predefined pod.
+type JobTemplate struct {
+	unversioned.TypeMeta `json:",inline"`
+	v1.ObjectMeta        `json:"metadata,omitempty"`
```
Comment (erictune) [inline pkg/apis/batch/v1/types.go]: use the same hyperlinks as PodTemplate comments

File: pkg/apis/extensions/types.go [hunk 2]
```diff
@@ -522,6 +522,24 @@ type JobList struct {
 	Items []Job `json:"items"`
 }
 
+// JobTemplate describes a template for creating copies of a predefined pod.
+type JobTemplate struct {
+	unversioned.TypeMeta `json:",inline"`
+	api.ObjectMeta       `json:"metadata,omitempty"`
```
Comment (erictune) [inline pkg/apis/extensions/types.go]: same here

**Comment by erictune** [comment]
I suggest we wait for after 1.2 to merge this.

**Comment by erictune** [comment]
So, this PR seems fine.  

I'm guessing the next PR is going to add "ScheduledJob".  I am not sure where that goes.
I don't think it can go into batch/v1 since `ScheduledJob` won't be `v1` on the first release.

I think what needs to happen is that we need to create a `batch/v2alpha1` which has `Job` and `ScheduledJob` in it, and that group needs to advance through `batch/v2beta1` and then `batch/v2`,
at which point `ScheduledJob` will be GA.  During that time, Jobs will be accessible through 3 or more endpoints.  It seems a bit complex, but I'm not sure what the alternative is.
@lavalamp @bgrant0607 @deads2k 

By the way, it makes it kind of an incentive to get Workflow in at the same time as ScheduledJob, to avoid a `v3alpha1`.

**Comment by soltysh** [comment]
> I suggest we wait for after 1.2 to merge this.

Yes, that was the point.

> I think what needs to happen is that we need to create a batch/v2alpha1 which has Job and ScheduledJob in it

Why can't we just use `Job` from `batch/v1` I'd personally stay away from creating Jobs in it. 

Other silly question, instead of messing `batch` sub-group, can't we use `extensions/v1beta1` or `extensions/v2beta1` as the incubator for new types, since the internal is already in place. And then move to appropriate group only when GA?

**Comment by soltysh** [comment]
@erictune added the urls you've asked for.

**Comment by sdminonne** [comment]
@erictune  @soltysh, I plan to submit a PR for workflow and for the moment it will be in `pkg/apis/extensions/v1beta1` but I'm OK to modify according to other needs/policy

**Comment by erictune** [comment]
@bgrant0607 @lavalamp we need to decide what to do in this case:

- we are introducing an alpha-version of a new Kind (e.g. ScheduledJob or Workflow) 
- we know already what the API group for the new kind it will eventually be (e.g. Batch)
- there is already something in the eventual destination group (e.g. Job is already in Batch/v1).

Considerations:
- extensions/v1beta1 is where we have been putting stuff, so it is easy/consistent to put new things there
- but... extensions/v1beta1 was a hack for when we didn't have api groups working in general.  we should stop using it now?
- the internal for Job is in extensions/v1beta1, and so putting things that use it in extensions/v1beta1 is convenient
- but... we have a plan to remove Job from there in version 1.4, and we would not be able to migrate ScheduledJob or Workflow out of extensions on the same schedule
- assuming ScheduledJob or Workflow went into extensions first, and then was ready for GA, where does it go for GA?  Batch/v1 or Batch/v2?  The former makes sense because there is not any change to Job planned, so no need to rev the group for that.  And it is less work.  The latter makes sense since we once said that we would not add new things to an API group once it had been formed, so that users could use the presence of the api group to determine the presence of all enclosing kinds.

**Comment by bgrant0607** [comment]
@erictune If there's nothing here that needs to be decided for 1.2, I'll have to look at this later.

**Comment by erictune** [comment]
no 1.2 issues.

**Comment by lavalamp** [comment]
Ping me again after 1.2 and I'll have lots to say. :)

**Comment by sdminonne** [comment]
@soltysh ping :)

**Comment by lavalamp** [comment]
I'm paying attention to PRs again. I will come back to this one once I triage my entire queue.

**Comment by sdminonne** [comment]
Thanks

**Comment by soltysh** [comment]
@lavalamp yeah, that will be needed for the scheduled jobs I'm hoping to start next week.

File: pkg/apis/batch/v1/types.go [hunk 3]
```diff
@@ -48,6 +48,29 @@ type JobList struct {
...
+	// Template defines jobs that will be created from this template
+	// http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+	Template JobTemplateSpec `json:"template,omitempty"`
+}
+
+// JobTemplateSpec describes the data a Job should have when created from a template
+type JobTemplateSpec struct {
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	v1.ObjectMeta `json:"metadata,omitempty"`
```
Comment (lavalamp) [inline pkg/apis/batch/v1/types.go]: You should not need this except in top-level objects.

File: pkg/apis/batch/v1/types.go [hunk 4]
```diff
@@ -48,6 +48,29 @@ type JobList struct {
...
+// JobTemplate describes a template for creating copies of a predefined pod.
+type JobTemplate struct {
+	unversioned.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	v1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Template defines jobs that will be created from this template
+	// http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+	Template JobTemplateSpec `json:"template,omitempty"`
```
Comment (lavalamp) [inline pkg/apis/batch/v1/types.go]: Do you need the JobTemplateSpec at all? Why not just JobSpec?

File: pkg/apis/extensions/types.go [hunk 5]
```diff
@@ -525,6 +525,29 @@ type JobList struct {
 	Items []Job `json:"items"`
 }
 
+// JobTemplate describes a template for creating copies of a predefined pod.
+type JobTemplate struct {
```
Comment (lavalamp) [inline pkg/apis/extensions/types.go]: Don't add JobTemplate here? Do you really need it in both places?

File: pkg/apis/extensions/v1beta1/types.go [hunk 6]
```diff
@@ -528,6 +528,24 @@ type JobList struct {
 	Items []Job `json:"items"`
 }
 
+// JobTemplate describes a template for creating copies of a predefined pod.
+type JobTemplate struct {
```
Comment (lavalamp) [inline pkg/apis/extensions/v1beta1/types.go]: Sorry, I meant don't add it here.

File: pkg/apis/extensions/types.go (hunk 5 above)
Comment (lavalamp) [inline pkg/apis/extensions/types.go]: Sorry, this one I understand why you're adding.

File: pkg/apis/extensions/v1beta1/types.go (hunk 6 above)
Comment (soltysh) [inline pkg/apis/extensions/v1beta1/types.go]: Yeah, that makes sense, since we'll be deprecating this endpoint soon.

File: pkg/apis/batch/v1/types.go (hunk 3 above)
Comment (soltysh) [inline pkg/apis/batch/v1/types.go]: This is similar to `PodTemplateSpec` to allow us define created Job's metadata fields.

File: pkg/apis/batch/v1/types.go (hunk 4 above)
Comment (soltysh) [inline pkg/apis/batch/v1/types.go]: see prev comment for explanation

File: pkg/apis/batch/v1/types.go (hunk 3 above)
Comment (soltysh) [inline pkg/apis/batch/v1/types.go]: I'll update the comment to give my reasoning.

**Comment by soltysh** [comment]
@lavalamp removed from v1beta1 and updated the comments, ptal

File: pkg/apis/batch/v1/types.go (hunk 4 above)
Comment (lavalamp) [inline pkg/apis/batch/v1/types.go]: Ah, of course.

File: pkg/apis/extensions/types.go (hunk 5 above)
Comment (lavalamp) [inline pkg/apis/extensions/types.go]: (Of course this should get moved to pkg/apis/batch/types.go in the near future, but I guess it's ok here now)

**Comment by lavalamp** [comment]
@k8s-bot test this issue: #23365

OK, this LGTM.

**Comment by lavalamp** [comment]
Welll... I don't know, actually I'm a bit nervous just adding a thing to a GA api. Maybe we should make batch/v2alpha1?

File: pkg/apis/extensions/types.go (hunk 5 above)
Comment (soltysh) [inline pkg/apis/extensions/types.go]: That will be done in a separate PR, this PR was initiated by @sdminonne request and the move was mentioned just recently.

**Comment by soltysh** [comment]
> Welll... I don't know, actually I'm a bit nervous just adding a thing to a GA api. Maybe we should make batch/v2alpha1?

@erictune @bgrant0607 opinions?

**Comment by bgrant0607** [comment]
v2alpha1 SGTM

File: pkg/apis/batch/v1/types.go [hunk 7]
```diff
@@ -48,6 +48,29 @@ type JobList struct {
...
+}
+
+// JobTemplateSpec describes the data a Job should have when created from a template
+type JobTemplateSpec struct {
+	// Standard object's metadata of the jobs created from this template.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	v1.ObjectMeta `json:"metadata,omitempty"`
```
Comment (bgrant0607) [inline pkg/apis/batch/v1/types.go]: This problem exists in all existing controllers also, but we should really create a stripped-down ObjectMeta to use in templates that only contains the relevant fields. The other fields are a liability to evolution, since the fields will actually be decoded and persisted, and make our documentation misleading and our schema validation incorrect.

**Comment by lavalamp** [comment]
v2alpha1 means:

* make the batch/v2alpha1 directory
* copy everything from v1 to it
* add the new type

Let me know if you encounter any difficulty getting it to work, we haven't
had to do this in a while but we definitely want to do it more in the
future so we need to streamline the process.

**Comment by soltysh** [comment]
@lavalamp will do, thx.

**Comment by erictune** [comment]
v2alpha1 plan is okay with me.

File: pkg/apis/batch/v1/types.go (hunk 7 above)
Comment (soltysh) [inline pkg/apis/batch/v1/types.go]: Moved to https://github.com/kubernetes/kubernetes/issues/23475.

**Comment by soltysh** [comment]
@lavalamp updated, ptal
The only problem I have is locally failing `make verify` with this error:
```
Verifying hack/../hack/verify-godep-licenses.sh
Checking for 'Godeps/' changes against '/job_template'
fatal: Not a valid object name /job_template
FAILED   hack/../hack/verify-godep-licenses.sh	0s
Verifying hack/../hack/verify-godeps.sh
Checking for 'Godeps/' changes against '/job_template'
fatal: Not a valid object name /job_template
FAILED   hack/../hack/verify-godeps.sh	0s
```
@erictune any ideas?

**Comment by soltysh** [comment]
@erictune sorry, I've meant @eparis ^

**Comment by lavalamp** [comment]
I think something about your local git clone is breaking the script. Could it be that it doesn't like branch names with underscores? It would be astonishing if you're the only person to do that.

**Comment by lavalamp** [comment]
The problem that the unit tests have is `pkg/apis/batch/v2alpha1/register.go:21: imported and not used: "k8s.io/kubernetes/pkg/api/v1"`

File: pkg/apis/batch/v2alpha1/register.go [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+var SchemeGroupVersion = unversioned.GroupVersion{Group: batch.GroupName, Version: "v2alpha1"}
+
+func AddToScheme(scheme *runtime.Scheme) {
+	addKnownTypes(scheme)
+}
+
+// Adds the list of known types to api.Scheme.
+func addKnownTypes(scheme *runtime.Scheme) {
+	scheme.AddKnownTypes(SchemeGroupVersion,
+		&JobTemplate{},
```
Comment (lavalamp) [inline pkg/apis/batch/v2alpha1/register.go]: You need to add all of the types here.

File: pkg/apis/batch/v2alpha1/types.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+*/
+
+package v2alpha1
+
+import (
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/api/v1"
+	batchv1 "k8s.io/kubernetes/pkg/apis/batch/v1"
+)
+
```
Comment (lavalamp) [inline pkg/apis/batch/v2alpha1/types.go]: You need to copy all of the types from v1 here. Each version needs to be a complete set.

File: pkg/apis/batch/v2alpha1/register.go [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+		&batchv1.Job{},
+		&batchv1.JobList{},
```
Comment (lavalamp) [inline pkg/apis/batch/v2alpha1/register.go]: Sorry, don't just reference from batch/v1, actually copy the types into v2alpha1/types.go.

File: pkg/apis/batch/v1/register.go [hunk 11]
```diff
@@ -19,14 +19,12 @@ package v1
 import (
 	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/apis/batch"
```
Comment (lavalamp) [inline pkg/apis/batch/v1/register.go]: Do not introduce this dependency!

File: pkg/apis/batch/v1/register.go [hunk 12]
```diff
@@ -19,14 +19,12 @@ package v1
...
 	"k8s.io/kubernetes/pkg/runtime"
 )
 
-// GroupName is the group name use in this package
-const GroupName = "batch"
```
Comment (lavalamp) [inline pkg/apis/batch/v1/register.go]: It is OK to repeat this because batch/v1 needs to be importable without batch/.

File: hack/test-go.sh [hunk 13]
```diff
@@ -58,7 +58,7 @@ KUBE_GOVERALLS_BIN=${KUBE_GOVERALLS_BIN:-}
 # Lists of API Versions of each groups that should be tested, groups are
 # separated by comma, lists are separated by semicolon. e.g.,
 # "v1,compute/v1alpha1,experimental/v1alpha2;v1,compute/v2,experimental/v1alpha3"
-KUBE_TEST_API_VERSIONS=${KUBE_TEST_API_VERSIONS:-"v1,extensions/v1beta1,metrics/v1alpha1;v1,autoscaling/v1,batch/v1,extensions/v1beta1,metrics/v1alpha1"}
+KUBE_TEST_API_VERSIONS=${KUBE_TEST_API_VERSIONS:-"v1,extensions/v1beta1,metrics/v1alpha1;v1,autoscaling/v1,batch/v1,batch/v2alpha1,extensions/v1beta1,metrics/v1alpha1"}
```
Comment (lavalamp) [inline hack/test-go.sh]: This will slow our tests down even more, but you may need to start a new group, I think this is selecting a version from each group so it will confuse the system to have a group in it twice?

e.g. append something like `;v1,autoscaling/v1,batch/v2alpha1,extensions/v1beta1`

File: hack/test-go.sh [hunk 14]
```diff
@@ -326,7 +326,7 @@ for (( i=0, j=0; ; )); do
   # KUBE_TEST_API sets the version of each group to be tested. KUBE_API_VERSIONS
   # register the groups/versions as supported by k8s. So KUBE_API_VERSIONS
   # needs to be the superset of KUBE_TEST_API.
-  KUBE_TEST_API="${apiVersion}" KUBE_API_VERSIONS="v1,autoscaling/v1,batch/v1,extensions/v1beta1,componentconfig/v1alpha1,metrics/v1alpha1,authorization.k8s.io/v1beta1" ETCD_PREFIX=${etcdPrefix} runTests "$@"
+  KUBE_TEST_API="${apiVersion}" KUBE_API_VERSIONS="v1,autoscaling/v1,batch/v1,batch/v2alpha1,extensions/v1beta1,componentconfig/v1alpha1,metrics/v1alpha1,authorization.k8s.io/v1beta1" ETCD_PREFIX=${etcdPrefix} runTests "$@"
```
Comment (lavalamp) [inline hack/test-go.sh]: I think it's ok to have both in this list. @nikhiljindal to confirm.

**Comment by soltysh** [comment]
I've copied entire `types.go` into `v2alpha1`, but I'm still missing some generations apparently, since tests are failing. Working on it...

**Comment by soltysh** [comment]
@lavalamp the linked issue might be fixed with this PR https://github.com/kubernetes/kubernetes/pull/23208 by @deads2k. It would definitely be beneficial to have that one first and only then come back to this one. I'll be looking into that other PR, appreciate your time as well on that PR.

**Comment by erictune** [comment]
@lavalamp @soltysh Are you saying that this PR is blocked on  #23208?

**Comment by deads2k** [comment]
> @lavalamp @soltysh Are you saying that this PR is blocked on #23208?

From what I've gathered, its a question of enablement.  It's currently hard to reason about which versions of which APIs are enabled, which storage encoding they're using, and which storage destination is in play.  Being unable to reason about that is making it hard to work out why multiple versions of the same API group don't seem to be starting correctly

File: hack/test-go.sh (hunk 14 above)
Comment (nikhiljindal) [inline hack/test-go.sh]: Yes should be fine.
We will get to know that its not fine if the tests fail :)

**Comment by erictune** [comment]
@lavalamp @soltysh how is this going?  Can I do anything to unblock it or otherwise help progress on ScheduledJob?

**Comment by soltysh** [comment]
@erictune sorry, was busy closing openshift 1.2. I'm back to the topic. I'm updating the PR and will ask for help where needed. What's currently block is the linked issue. I'll try to rebase my PR on top of @deads2k changes.

**Comment by soltysh** [comment]
I've rebased and updated the PR with remaining conversions fixes (see last commit for that). The last problem I'm struggling with is clients, iow. this error in unit tests:
```
--- FAIL: TestListJob (0.00s)
...
received: "/apis/batch/v1/jobs", expected: "/apis/batch/v2alpha1/jobs"
```
From looking into client code [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/jobs.go#L106) and [here](https://github.com/kubernetes/kubernetes/blob/31de62216dea226f0553e698da38000b73684d49/pkg/client/unversioned/batch.go#L36) our unversioned client is not that much unversioned. The question is how to add `batch/v2alpha1` there?
@erictune @lavalamp ideas?

**Comment by lavalamp** [comment]
I added a line to the test framework to let you fix that error, IIRC. ... Yeah, it was for that test, too.

This resource/version/group: https://github.com/kubernetes/kubernetes/blob/dae5ac482861382e18b1e7b2943b1b7f333c6a2a/pkg/client/unversioned/jobs_test.go#L106

Needs to match the path constructed here: https://github.com/kubernetes/kubernetes/blob/dae5ac482861382e18b1e7b2943b1b7f333c6a2a/pkg/client/unversioned/jobs_test.go#L88

Right now the test is called twice. Now you should call it three times, for extensions/v1beta1, for batch/v1, and batch/v2alpha1.

**Comment by soltysh** [comment]
@lavalamp thanks will update shortly

**Comment by soltysh** [comment]
I've rebase this PR on top of #23493 since it's almost in. One remaining thing, I'm working on right now, is failing tests in `pkg/master`. Hopefully I'll be able to nail that down tomorrow.

**Comment by soltysh** [comment]
@caesarxuchao I found one leftover from #23493, please have a look at [first commit](https://github.com/kubernetes/kubernetes/pull/21675/commits/102b94a33eee636e0b37e6784ffcfd7b155ccf56).

@lavalamp I've slightly extended `testapi.go` (see [last commit](https://github.com/kubernetes/kubernetes/pull/21675/commits/dca4d75c0b793c17923a9461319de19844c5eb95)) to allow having multiple versions at once, this works, but to have fully testable multiple versions (the one in `pkg/client/unversioned/jobs_test.go` you've pointed me to) I need to modify `simple_testclient.go` to allow choosing which version to go with, see this [TODO](https://github.com/kubernetes/kubernetes/blob/d609e6cc98ab27acbe7a7e4101a1d9fe65e354ec/pkg/client/unversioned/testclient/simple/simple_testclient.go#L90-L91) (@caesarxuchao any idea how to fix it?). 
I've started working on the fixes for that by extending [ResourcePath](https://github.com/kubernetes/kubernetes/blob/d609e6cc98ab27acbe7a7e4101a1d9fe65e354ec/pkg/api/testapi/testapi.go#L241) and [ResourcePathWithPrefix](https://github.com/kubernetes/kubernetes/blob/d609e6cc98ab27acbe7a7e4101a1d9fe65e354ec/pkg/api/testapi/testapi.go#L211) to pass which version you're specifically want to go with, but I'm not sure that's the right approach, I'd rather submit a separate PR addressing that problem, if you don't mind. This PR seems to be already big enough.

**Comment by caesarxuchao** [comment]
> @caesarxuchao I found one leftover from #23493, please have a look at first commit.

That LGTM.

> I need to modify simple_testclient.go to allow choosing which version to go with, see this TODO (@caesarxuchao any idea how to fix it?). 

IIUC, you want to test multiple versions of the batch group? I think we need to support that in the real client as well, not in simple_testclient.go. IIRC, what we used to do was initializing multiple clients for each version.

**Comment by lavalamp** [comment]
Re: testing multiple versions, there are some ENV vars in the test scripts
that let you run the tests multiple times with various groups turned on. We
have two iterations right now, may need to add a 3rd.

**Comment by caesarxuchao** [comment]
The ENV lavalamp is talking about: https://github.com/kubernetes/kubernetes/blob/945daaae91722d0fe59adb689789c20966544ece/hack/test-go.sh#L61

**Comment by soltysh** [comment]
@caesarxuchao make sense, I was a little bit confused after talking to @lavalamp yesterday when he mentioned having all versions (iow. both `v1` and `v2alpha1`) enabled at the same time. I'll update the PR with that limitation in mind, at least for now.

**Comment by lavalamp** [comment]
sorry for not being clear-- I'm kinda buried right now. If it's easier,
feel free to just turn on all three at the same time, using that env var.

**Comment by soltysh** [comment]
No worries, I'll revisit the approach tomorrow morning and would like to push final version tomorrow. I hope with all the bits around versioning settled I'll be able to do so.

**Comment by soltysh** [comment]
@erictune @lavalamp this is ready for final review. 
A couple of takeaways that needs addressing, but in separate PRs:
- currently we enable all versions at once from a single group, this is because our `install.go` is at a group level and is installing the entire group with all its versions, it needs to be more granular, imho. This is why I've added `batch/v2alpha1` to the same testing group as `batch/v1`.
- we need multiversion client support, to enable testing different versions at once, not possible yet, eg. in `jobs_test.go` only `extensions/v1beta1` and `batch/v1` is being tested. Some small part were addressed to allow `master_test.go` to pass, see the last commit (enable batch/v2alpha1 tests).

**Comment by soltysh** [comment]
I was missing newline at the end of `pkg/apis/batch/v2alpha1/generated.proto`.

**Comment by erictune** [comment]
@lavalamp this PR is at that magical moment when all its tests are passing and it does not need a rebase.  It just needs an LGTM.  If you could LGTM soon that would be awesome.

File: pkg/master/master.go [hunk 15]
```diff
@@ -316,7 +319,8 @@ func (m *Master) InstallAPIs(c *Config) {
 		apiGroupInfo := genericapiserver.APIGroupInfo{
 			GroupMeta: *batchGroupMeta,
 			VersionedResourcesStorageMap: map[string]map[string]rest.Storage{
-				"v1": batchResources,
+				"v1":       batchv1Resources,
+				"v2alpha1": batchv2alpha1Resources,
```
Comment (lavalamp) [inline pkg/master/master.go]: I think you should probably check that the version is enabled before adding to this map, so like construct the map here and have two `if c.APIResourceConfigSource.AnyResourcesForVersionEnabled(batchapiv2alpha1.SchemeGroupVersion) { apiGroupInfo.VersionedResourcesStorageMap["v2alpha1"] = batchv2alpha1Resources }` statements afterwards.

I don't think it's that big of a deal but without this there's no way to e.g. turn the alpha version off and leave the other version on.

File: pkg/apis/batch/v2alpha1/conversion.go [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+	} else {
+		out.Completions = nil
+	}
+	if in.ActiveDeadlineSeconds != nil {
+		out.ActiveDeadlineSeconds = new(int64)
+		*out.ActiveDeadlineSeconds = *in.ActiveDeadlineSeconds
+	} else {
+		out.ActiveDeadlineSeconds = nil
+	}
+	// unable to generate simple pointer conversion for unversioned.LabelSelector -> v2alpha1.LabelSelector
```
Comment (lavalamp) [inline pkg/apis/batch/v2alpha1/conversion.go]: I wasn't going to say anything, but since I already asked for a change, does it work if you just provide the Convert_unversioned_LabelSelector_To_v2alpha1_LabelSelector function? I think the autogenerated conversion functions should call that if you provide it, and then you wouldn't need this function.

File: pkg/api/testapi/testapi.go [hunk 17]
```diff
@@ -55,9 +55,10 @@ var (
 )
 
 type TestGroup struct {
-	externalGroupVersion unversioned.GroupVersion
-	internalGroupVersion unversioned.GroupVersion
-	internalTypes        map[string]reflect.Type
+	// the first element in the group is meant to be the preferred version
```
Comment (lavalamp) [inline pkg/api/testapi/testapi.go]: Thanks for dealing with this file. I think it needs some help in the future but I won't ask you do to that.

**Comment by lavalamp** [comment]
It looks great, just one minor thing.

File: pkg/master/master.go (hunk 15 above)
Comment (soltysh) [inline pkg/master/master.go]: Done.

File: pkg/api/testapi/testapi.go (hunk 17 above)
Comment (soltysh) [inline pkg/api/testapi/testapi.go]: Yes it does :/. My work is just a tip of an iceberg we need to dig into to get multiple versions fully testable.

File: pkg/apis/batch/v2alpha1/conversion.go (hunk 16 above)
Comment (soltysh) [inline pkg/apis/batch/v2alpha1/conversion.go]: I've tried it, but although it didn't generate `Convert_unversioned_LabelSelector_To_v2alpha1_LabelSelector` (or the other side). It still generate `autoConvert_unversioned_LabelSelector_To_v2alpha1_LabelSelector`. Which I've decided will be unnecessary code duplication. I'll leave this as is in that case.

**Comment by soltysh** [comment]
@lavalamp I've rebased the PR and addressed that comment regarding turning on the storage. Additionally, I've included the swagger spec leftovers I've found when working on scheduled job (see [this commit](https://github.com/kubernetes/kubernetes/pull/24970/commits/613f3e293c849fa128bd3cee3236464ae4b87a4e)) for that particular changes I've included here. ptal

**Comment by lavalamp** [comment]
LGTM, thanks!!

**Comment by erictune** [comment]
This is so close to the head of the submit queue.  🍀 🙏

**Comment by soltysh** [comment]
Rebased. @erictune || @lavalamp mind re-adding the lgtm label.

**Comment by soltysh** [comment]
Sorry, forgot to run `hack/update-all.sh`.

**Comment by soltysh** [comment]
Rebased, I didn't forget to run `hack/update-all.sh` this time, so hopefully should be a green one. I've also updated `defaults.go` according to recent changes. 
@erictune || @lavalamp mind re-adding the lgtm label.

**Comment by soltysh** [comment]
@erictune || @lavalamp mind re-adding the lgtm label, I'd like to avoid rebasing once again.

**Comment by soltysh** [comment]
@k8s-bot test this github issue: #24211

**Comment by soltysh** [comment]
:tada:

**Comment by erictune** [comment]
👍
